tests/unit/dataactvalidator/test_c9_award_financial.py

```diff
@@ -1,37 +1,151 @@
-    # Test with both fain and uri
-    award_fin = AwardFinancialFactory()
-    afa = AwardFinancialAssistanceFactory(fain = award_fin.fain, uri = award_fin.uri, federal_action_obligation = 345)
-    # Test with zero federal action obligation
-    afa_zero = AwardFinancialAssistanceFactory(federal_action_obligation = 0)
-
-    assert number_of_errors(_FILE, database, models=[award_fin_fain, award_fin_uri, afa_fain, afa_uri, afa_zero,
-                                                     award_fin, afa]) == 0
-
-def test_failure(database):
-    """ Test fain and uri present in D2 but not in C """
-    award_fin = AwardFinancialFactory(fain = None, uri = None)
-    afa_fain = AwardFinancialAssistanceFactory(uri = None, federal_action_obligation = 123)
-    afa_uri = AwardFinancialAssistanceFactory(fain = None, federal_action_obligation = 234)
-    # Test with both fain and uri
-    afa = AwardFinancialAssistanceFactory(federal_action_obligation = 345)
-
-    assert number_of_errors(_FILE, database, models=[award_fin, afa_fain, afa_uri, afa]) == 3
+def test_equal_fain(database):
+    """Tests that File D2 (award financial assistance) fain matches
+    File C (award financial) fain."""
+    tas = _TAS
+    afa = AwardFinancialAssistanceFactory(
+        tas=tas, fain='abc', uri=None, federal_action_obligation='1', original_loan_subsidy_cost='1')
+    af = AwardFinancialFactory(
+        tas=tas, submisson_id=afa.submission_id, fain=afa.fain, uri=None)
+
+    errors = number_of_errors(_FILE, database, models=[afa, af])
+    assert errors == 0
+
+
+def test_equal_uri(database):
+    """Tests that File D2 (award financial assistance) uri matches
+    File C (award financial) uri."""
+    tas = _TAS
+    afa = AwardFinancialAssistanceFactory(
+        tas=tas, fain=None, uri='xyz', federal_action_obligation='1', original_loan_subsidy_cost='1')
+    af = AwardFinancialFactory(
+        tas=tas, submisson_id=afa.submission_id, fain=None, uri=afa.uri)
+
+    errors = number_of_errors(_FILE, database, models=[afa, af])
+    assert errors == 0
+
+
+def test_null_uri_fain(database):
+    """Tests File D2 (award financial assistance) and File C (award financial)
+    having NULL values for both fain and uri."""
+    tas = _TAS
+    afa = AwardFinancialAssistanceFactory(
+        tas=tas, fain=None, uri=None, federal_action_obligation='1', original_loan_subsidy_cost='1')
+    af = AwardFinancialFactory(
+        tas=tas, submisson_id=afa.submission_id, fain=None, uri=None)
+
+    errors = number_of_errors(_FILE, database, models=[afa, af])
+    assert errors == 0
+
+
+def test_both_fain_and_url_supplied(database):
+    """Tests File D2 (award financial assistance) having both uri and fain populated."""
+    tas = _TAS
+    afa = AwardFinancialAssistanceFactory(
+        tas=tas, fain='abc', uri='xyz', federal_action_obligation='1', original_loan_subsidy_cost='1')
+
+    errors = number_of_errors(_FILE, database, models=[afa])
+    assert errors == 1
```

Comment:
If C9 above is changed to apply the check to AF this will need to change as well